pyomo/contrib/cp/scheduling_expr/step_function_expressions.py

```diff
@@ -0,0 +1,282 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.contrib.cp.interval_var import (
+    IntervalVar, IntervalVarData, IntervalVarStartTime, IntervalVarEndTime)
+from pyomo.core.base.component import Component
+from pyomo.core.expr.logical_expr import BooleanExpression
+from pyomo.core.pyomoobject import PyomoObject
+
+def _generate_sum_expression(_self, _other):
+    # We check both because we call this function for the reverse operation as
+    # well.
+    if isinstance(_other, StepFunction) and isinstance(_self, StepFunction):
+        if _self.nargs() == len(_self._args_):
+            _self._args_.extend(_other.args)
+            return CumulativeFunction(_self._args_, nargs=len(_self._args_))
+        else:
+            # we have to clone the list of _args_
+            return CumulativeFunction(_self.args + _other.args)
+    else:
+        raise TypeError("Cannot add object of class %s to object of class "
+                        "%s" % (_other.__class__, _self.__class__))
+
+def _generate_difference_expression(_self, _other):
+    # We check both because we call this function for the reverse operation as
+    # well.
+    if isinstance(_other, StepFunction) and isinstance(_self, StepFunction):
+        if _self.nargs() == len(_self._args_):
+            _self._args_.extend([NegatedStepFunction(a) for a in _other.args])
+            return CumulativeFunction(_self._args_, nargs=len(_self._args_))
+        else:
+            # we have to clone the list of _args_
+            return CumulativeFunction(_self.args + [NegatedStepFunction(a) for a
+                                                    in _other.args])
+    else:
+        raise TypeError("Cannot subtract object of class %s from object of "
+                        "class %s" % (_other.__class__, _self.__class__))
+
+class StepFunction(PyomoObject):
```

Comment:
I was hiding it from the expression system so that it would be treated as a leaf node, but I can do that explicitly in the writer. I think it should inherit from `ExpressionBase`